lib/client/api.go

```diff
@@ -1314,95 +1316,116 @@ func (tc *TeleportClient) Join(ctx context.Context, namespace string, sessionID
 
 // Play replays the recorded session
 func (tc *TeleportClient) Play(ctx context.Context, namespace, sessionID string) (err error) {
-	if namespace == "" {
-		return trace.BadParameter(auth.MissingNamespaceError)
-	}
-	sid, err := session.ParseID(sessionID)
-	if err != nil {
-		return fmt.Errorf("'%v' is not a valid session ID (must be GUID)", sid)
-	}
-	// connect to the auth server (site) who made the recording
-	proxyClient, err := tc.ConnectToProxy(ctx)
-	if err != nil {
-		return trace.Wrap(err)
-	}
-	defer proxyClient.Close()
+	var sessionEvents []events.EventFields
+	var stream []byte
+	// check if user used file as argument
+	if path.Ext(sessionID) == ".tar" {
+		tarFile, err := os.Open(sessionID)
+		if err != nil {
+			return trace.ConvertSystemError(err)
+		}
+		defer tarFile.Close()
+		protoReader := events.NewProtoReader(tarFile)
+		file := filepath.Base(sessionID)
+		sid := strings.TrimSuffix(file, ".tar")
```

Comment:
I'd try to encapsulate all domain-level knowledge in specific packages. In this case, it seems that the input file has a sessionID embedded into it - there should be a function to extract this detail in the same package that generated the filename.
Similar, low-level handling of chunks and events is all implementation detail of the session playback format which would be hidden inside the `lib/events` package. I would approach this by encapsulating this knowledge in a value that `WriteForPlayback` returns instead of using raw filepaths - this goes for reading the events and chunks - all not relevant here.

Reply:
I think a good way to implement is to still have `WriteForPlayback` still return the paths (that just seems fitting), but then have 2 other functions in events that return `[]EventFields` and `[]byte` that is the stream like below. And these methods would be called to obtain the input the player needs. 

`func SessionEvents(eventsPath string) ([]EventFields, error) `

`func SessionChunks(chunksPath string) ([]byte, error) `